### PR TITLE
Dynamo comptime, with public ComptimeContext API

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90983

This PR adds `@comptime`, a decorator that causes a given function to be executed at compile time when Dynamo is symbolically evaluating their program. To query the Dynamo state, we offer a public ComptimeContext API which provides a limited set of APIs for querying Dynamo's internal state. We intend for users to use this API and plan to keep it stable. Here are some things you can do with it:

* You want to breakpoint Dynamo compilation when it starts processing a particular line of user code: give comptime a function that calls breakpoint
* You want to manually induce a graph break for testing purposes; give comptime a function that calls unimplemented
* You want to perform a debug print, but you don't want to induce a graph break; give comptime a function that prints.
* You can print what the symbolic locals at a given point in time are.
* You can print out the partial graph the Dynamo had traced at this point.
* (My original motivating use case.) You want to add some facts to the shape env, so that a guard evaluation on an unbacked SymInt doesn't error with data-dependent. Even if you don't know what the final user API for this should be, with comptime you can hack out something quick and dirty. (This is not in this PR, as it depends on some other in flight PRs.)

Check out the tests to see examples of comptime in action.

In short, comptime is a very powerful debugging tool that lets you drop into Dynamo from user code, without having to manually jerry-rig pdb inside Dynamo to trigger after N calls.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Comment by voznesenskym** [comment]
Brief thoughts, discussed offline somewhat. 

Philosophically, I think this is a great tool. This is v. similar to what I mentioned in the NFC chat (cc @Chillee @zou3519) w/r/t to giving users a VariableTracker, hooked into the builder, to populate as a compile time alternative to help interpret their types instead of just/only using @allow_in_graph. 

Implementation wise - exposing tx, output_graph, and other core internals is going to be an extremely difficult contract to maintain. It essentially broadens the entire surface area of what we have to consider as "user facing" - to the entirety of dynamo. If we do go ahead with this, I think we need to either:

A) Expose a clear subset API like `OutputGraphPublic`, `InstructionTranslatorPublic`, where `InstructionTranslator` is now `InstructionTranslator(InstructionTranslatorPublic)` and type-erase that down to just `InstructionTranslatorPublic` when we pass it back:

```
class InstructionTranslatorPublic:

@property
def output() -> OutputGraphPublic:

class InstructionTranslator(InstructionTranslatorPublic):

@property
def _output() -> OutputGraph
```

This will be annoying to maintain, and add a lot of class structures, but will also signal to users what is and is not a real API, as well as force us to think about what is, and is not, okay to expose or contractualize

B) Do what you did in the PR, but give it extremely clear signaling around unsafety, like `comptime_not_back_compat_ever` or `comptime_unsafe` or something like that, and always make sure the contract is `*args/**kwargs` so even new args or removal of old ones doesnt break contract.

**Comment by mlazos** [comment]
I'm really happy with this idea and would definitely use this, mainly as a debugging tool. I think your suggestion about having this be an underlying API which we narrowly expose to users with small public facing APIs is a good idea to address the concern of exposing all of dynamo as a public contract.

**Review by jansel — COMMENTED** [review]
I'm ok with this idea as a debugging tool, though I wouldn't want to maintain it as a BC surface.  So perhaps add some sort of giant warning on it.

**Comment by ezyang** [comment]
This PR has been updated with a proposal API for the public bindings that allow you to access Dynamo internal state. I took great pains to reduce the number of commitments we make to future users, while making the API flexible enough to allow you to do most things you might want to do when debugging. Check it out!

File: test/dynamo/test_comptime.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        global FILE
+        FILE = StringIO()
+
+        @torch._dynamo.optimize("eager")
+        def f(x):
+            y = x * 2
+
+            @comptime
+            def _(ctx):
+                ctx.print_disas(file=FILE)
```
Comment (jansel) [inline test/dynamo/test_comptime.py:51]: Perhaps we should expose some of these through a more compact API like `dynamo.debug.print_disas()` (which would just be this same function) -- I'd expect a lot of common cases will just be calling a single one of these functions.

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/comptime.py [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+        are used.
+        """
+        # TODO: improve print format, current guard format is extremely
+        # verbose
+        print(
+            "\n".join(f"-{str(guard)}" for guard in sorted(self.__tx.output.guards)),
+            file=file,
+        )
+
+    def _i_will_not_complain_if_bc_breaks_InstructionTranslator(self):
```
[the remark this answers is not in the document]
Comment (ezyang) [reply torch/_dynamo/comptime.py:212]: Ah but you see, the glorious thing is that double underscore prefixed identifiers in Python get name mangled, so you'd have to write `ctx._ComptimeContext__tx` anyway! Take that! (Well, I guess the mangled name is still shorter though haha)

File: test/dynamo/test_comptime.py (hunk 1 above)
Comment (ezyang) [reply test/dynamo/test_comptime.py:51]: The one subtlety is that we would have to make all of these functions special forms in Dynamo, in the same way `comptime` has special handling. We can still do that though. Another possibility is to just define the function as

```
def print_disas(ctx):
    ctx.print_disas()
```

and then you can simply say `comptime(torch._dynamo.debug.print_disas)` to do the single line, without having to introduce other special forms (but you'd still have to import modules for two identifiers.) These are both easy to implement so it's mostly a matter of what people prefer.

Comment (ezyang) [reply test/dynamo/test_comptime.py:51]: Actually, I guess if you don't mind introducing a stack frame (which does matter in a few cases), you can just use functional abstraction. I'll go add these.

**Comment by ezyang** [comment]
Updated:
- API now supports stacklevel so you can parent outer InstructionTranslator
- Convenience API comptime.print_X() now exists, for compact debugging.

File: torch/_dynamo/variables/misc.py [hunk 3]
```diff
@@ -87,6 +88,61 @@ class UnknownVariable(VariableTracker):
...
+                code,
+                fn.f_globals,
+                fn.fn_name.as_python_constant(),
+                tuple(fn.defaults.items) if fn.defaults else None,
+                # We could automatically promote free variables into
+                # ComptimeVar but this is confusing if you access
+                # a free variable that we actually DO have the runtime
+                # value for
+                # tuple(make_cell(ComptimeVar(i)) for i in fn.closure.items)
+                tuple(),
```
Comment (ezyang) [inline torch/_dynamo/variables/misc.py:137]: So, automatically promoting free variables into ComptimeVar is hella confusing, so we shouldn't do it, but I also thought about automatically calling `as_python_constant` on free variables. The problem is I can't guarantee that object identity holds (e.g., you access a "constant" list, that list will actually get reconstructed) and then mutations don't work correctly. Also nonlocal handling with cells is a pain.
